### ISSUE TITLE
Runtime can generate multi-dispatch nodes

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4189]
This task is automatically imported from the old Task Issue Board and it was originally created by Dmitry Bushev.
Original issue is [here](https://www.pivotaltracker.com/story/show/180539669).
<hr />

### Why

**As Enso runtime**
**I want ** to generate multi-dispatch nodes
**So that ** universal conversion methods work

### Acceptance Criteria

```gherkin
Scenario: 
Given universal conversion method
When called
Then is invoked the correct method
```

**Notes:**

This is the interpreter part of the Universal conversions task.
- Implement the multi-dispatch-node
- Replace the ordinary method call node with the multi-dispatch node for .from calls. For reference, I added the generation of method calls here https://github.com/enso-org/enso/pull/3177
- Ensure that all standard features work with from calls (currying, oversaturated apps, default suspension)

## Comments:
**Edward Kmett** reports a new **STANDUP** for the provided date (2021-12-08):

**Today:** meeting with wdanilo to talk about next directions after universal conversions, started simpler/slower definition to unblock Dmitry It should be finished by 2021-12-18.

**Next Day:** Next day I will be working on the [##180539669 task](https://www.pivotaltracker.com/story/show/180539669). continue working on the same task (Enso Bot - Dec 10, 2021)
<hr />
**Edward Kmett** reports a new **STANDUP** for the provided date (2021-12-07):

**Today:** mostly meetings and interviews for compiler candidates. realigned to skip the polymorphic inline cache for now, this comes at a real performance hit (temporarily) but should let me unblock dmitry refactoring the standard library to use conversions significantly earlier rather than later, and we can revisit this based on how bad the tax is in practice. It should be finished by 2021-12-18.

**Next Day:** Next day I will be working on the [##180539669 task](https://www.pivotaltracker.com/story/show/180539669). continue working on the same task (Enso Bot - Dec 10, 2021)
<hr />
**Edward Kmett** reports a new **STANDUP** for the provided date (2021-12-03):

**Today:** rebasing on top of dmitry's patch #3177 It should be finished by 2021-12-18.

**Next Day:** Next day I will be working on the [##180539669 task](https://www.pivotaltracker.com/story/show/180539669). continue working on the same task (Enso Bot - Dec 10, 2021)
<hr />
**Edward Kmett** reports a new **STANDUP** for yesterday (2021-12-12):

**Today:** distinguished MultiDispatchRuntimeFunction and MultiDispatchFunction, started hooking up to Dmitry's front end code. Done in a larger time block to give me more focus time than allotted in 4 hour increments. It should be finished by 2021-12-17.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). More wiring.

**Additional Information:** 0 hours logged thursday or friday, worked on sunday instead to maximize focus time. (Enso Bot - Dec 13, 2021)
<hr />
**Edward Kmett** reports a new **STANDUP** for today (2021-12-13):

**Today:** split off a registerConversionMethod, started tackling invocation, Demo Day. It should be finished by 2021-12-17.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). More wiring. (Enso Bot - Dec 13, 2021)
<hr />
**Edward Kmett** reports a new **STANDUP** for the provided date (2021-11-29):

**Today:** exploring a proper approach for polymorphic inline caching of universal conversions. It should be finished by 2021-12-17.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Continuation (Enso Bot - Dec 14, 2021)
<hr />
**Edward Kmett** reports a new **STANDUP** for the provided date (2021-11-30):

**Today:** exploring a proper approach for polymorphic inline caching of universal conversions. It should be finished by 2021-12-17.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Continuation (Enso Bot - Dec 14, 2021)
<hr />
**Edward Kmett** reports a new **STANDUP** for the provided date (2021-12-01):

**Today:** exploring a proper approach for polymorphic inline caching of universal conversions. It should be finished by 2021-12-17.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Continuation (Enso Bot - Dec 14, 2021)
<hr />
**Edward Kmett** reports a new **STANDUP** for the provided date (2021-12-02):

**Today:** exploring a proper approach for polymorphic inline caching of universal conversions. It should be finished by 2021-12-17.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Continuation (Enso Bot - Dec 14, 2021)
<hr />
**Edward Kmett** reports a new **STANDUP** for yesterday (2021-12-14):

**Today:** trying to figure out a way to invoke the new conversion methods. Accepting a short term approach of just visiting them in order rather than topologically sorting and walking superclasses, as I want to get this to baseline usable ASAP. It should be finished by 2021-12-17.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). I may have reduced hours tomorrow, due to having contractors on site in my house. I currently intend to make up the balance on Friday. (Enso Bot - Dec 15, 2021)
<hr />
**Edward Kmett** reports a new **STANDUP** for today (2021-12-15):

**Today:** Had a frustrating experience trying to get graphs out of the truffle visualization tools to see why this isn't working. Will pick it up again tomorrow. It should be finished by 2021-12-17.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Will try to get the graphs to actually work successfully. (Enso Bot - Dec 16, 2021)
<hr />
**Edward Kmett** reports a new **🔴 DELAY** for yesterday (2021-12-16):

**Summary:** There is **2 days** delay in implementation of the [Runtime can generate multi-dispatch nodes (#180539669) task](https://www.pivotaltracker.com/story/show/180539669).
It **will cause 0 days delay** for the delivery of this sprint.

This is based on my previous assumption about the amount of remaining work being enough to carry me to Friday.

**Delay Cause:** I lost ~4 hours worth of work due to having to deal with contractors at my house.

**Possible solutions:** I'm going to try to finish this over the weekend. (Enso Bot - Dec 18, 2021)
<hr />
**Edward Kmett** reports a new **STANDUP** for yesterday (2021-12-16):

**Today:** imited availability due to contractors in the house. Will do more over the weekend. That will class as a 'delay', because it comes after Friday. Fought the graal graphing tool today and spent time trying to interpret what I could extract from it. It should be finished by 2021-12-19.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Cross-referencing these graphs with the code-base to fill in some gaps in the documentation about how invocation nodes work. Then figuring out how to finish this off. (Enso Bot - Dec 18, 2021)
<hr />
**Edward Kmett** reports a new **STANDUP** for today (2021-12-17):

**Today:** Started parsing through graal graphs. Day clipped slightly short due to working with contractors in the house, again. It should be finished by 2021-12-19.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Catching up on the 5 hours or so worth of work I missed this week, and trying to finally get this to a demoable state. (Enso Bot - Dec 18, 2021)
<hr />
**Edward Kmett** reports a new **🔴 DELAY** for yesterday (2021-12-19):

**Summary:** There is **6 days** delay in implementation of the [Runtime can generate multi-dispatch nodes (#180539669) task](https://www.pivotaltracker.com/story/show/180539669).
It **will cause 0 days delay** for the delivery of this sprint.

Current intention is to finish this by the end of the sprint, so Dec 26th. That would dovetail with the vvccccugjjbikuvltetgrbrultnjjvlgurbjcthkfdbhe fact that I'm out the following week, and that Dmitry is out the week after that.

**Delay Cause:** I spent the weekend chewing on this, and it looks like it is going to be trickier than I thought.

**Possible solutions:** Pushing to finish this over this sprint. (Enso Bot - Dec 20, 2021)
<hr />
**Edward Kmett** reports a new **🔴 DELAY** for yesterday (2021-12-19):

**Summary:** There is **1 days** delay in implementation of the [Runtime can generate multi-dispatch nodes (#180539669) task](https://www.pivotaltracker.com/story/show/180539669).
It **will cause 0 days delay** for the delivery of this sprint.

I can't add. I had intended this to end on Dec 26th, not the 25th.

**Delay Cause:** Math is hard.

**Possible solutions:** Learn math. (Enso Bot - Dec 21, 2021)
<hr />
**Edward Kmett** reports a new **STANDUP** for yesterday (2021-12-19):

**Today:** Realized that some simplifying assumptions I had made about the ability to strip pre-applied arguments from multiple dispatch nodes are actually not valid. It should be finished by 2021-12-26.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Probably adding back some logic for pre-applied arguments modified to suit multiple dispatch. This is going to require me to figure out some stuff I'd previously punted on around our calling conventions and other undocumented bits. (Enso Bot - Dec 21, 2021)
<hr />
**Edward Kmett** reports a new **STANDUP** for today (2021-12-20):

**Today:** Digging into our calling conventions and interpreting these graphs, started refactor to bring back pre-applied arguments. It should be finished by 2021-12-26.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Fixing up pre-applied arguments. (Enso Bot - Dec 21, 2021)
<hr />
**Edward Kmett** reports a new **STANDUP** for yesterday (2021-12-21):

**Today:** Worked on pre-applied arguments. It should be finished by 2021-12-26.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). More of the same (Enso Bot - Dec 23, 2021)
<hr />
**Edward Kmett** reports a new **STANDUP** for today (2021-12-22):

**Today:** Worked on pre-applied arguments. It should be finished by 2021-12-26.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Seeing if this elaborate rube goldberg machine actually works. (Enso Bot - Dec 23, 2021)
<hr />
**Edward Kmett** reports a new **🔴 DELAY** for yesterday (2022-01-03):

**Summary:** There is **21 days** delay in implementation of the [Runtime can generate multi-dispatch nodes (#180539669) task](https://www.pivotaltracker.com/story/show/180539669).
It **will cause 0 days delay** for the delivery of this sprint.

**Delay Cause:** This is significantly harder than I expected. A conversation with Marcin seems to have put me on the right track. But between vacations, christmas, new years, etc. and miscalibration before and wanting to actually have some tests in place, this should line up with my current expectations of when this will be complete and usable, when coupled with pairing with Marcin on some of this starting ~Friday. Will use this time to try to crib a more productive workflow from Marcin as well. (Enso Bot - Jan 5, 2022)
<hr />
**Edward Kmett** reports a new **STANDUP** for yesterday (2022-01-03):

**Today:** Started back in with fresh eyes in the new year. Accumulated a list of questions for Marcin. It should be finished by 2022-01-16.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Talking to Marcin. (Enso Bot - Jan 5, 2022)
<hr />
**Edward Kmett** reports a new **STANDUP** for today (2022-01-04):

**Today:** Talked to Marcin, worked up a plan of attack. It should be finished by 2022-01-16.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Applying some tweaks to the IrToTruffle.scala code to generate appropriate multiple dispatch application nodes, and maybe starting in on how to encapulate the 'from' logic in a "truffle library" so we can properly invoke it using polyglot, and not cut and paste the logic everywhere.

**Additional Information:** none (Enso Bot - Jan 5, 2022)
<hr />
**Marcin Kostrzewa** reports a new **STANDUP** for yesterday (2022-01-04):

**Today:** an onboarding call with Edward Kmett, some explanation of the relevant codebase bits and arrangements for future work It should be finished by 2022-01-16.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). a pair programming session with Ed (Enso Bot - Jan 5, 2022)
<hr />
**Edward Kmett** reports a new **STANDUP** for yesterday (2022-01-05):

**Today:** Researched truffle libraries and changes to graal since "my day". Spent some time reading the output of the compiler to try to tease out some of the rewrites and newer state machinery. It should be finished by 2022-01-16.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Interviewing Francois. More reading.

**Additional Information:** none (Enso Bot - Jan 7, 2022)
<hr />
**Edward Kmett** reports a new **STANDUP** for today (2022-01-06):

**Today:** Interviewed Francois. Will follow up with details separately. That hit in the middle of my day and didn't finish much of what I started otherwise. It should be finished by 2022-01-16.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). ~2 hours of pair programming with Marcin @ 8am Interviewing Hubert @ 10am Talking to wdanilo after.

**Additional Information:** none (Enso Bot - Jan 7, 2022)
<hr />
**Marcin Kostrzewa** reports a new **STANDUP** for today (2022-01-07):

**Today:** pair programming session with Edward, getting a significant chunk of the functionality off the ground and figuring out the full implementation plan. It should be finished by 2022-01-16.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). (tuesday) a pair programming session with Edward (Enso Bot - Jan 7, 2022)
<hr />
**Edward Kmett** reports a new **STANDUP** for today (2022-01-07):

**Today:** Pair programmed with Marcin. Moved code out to compiler/wip/eak/conversions branch. Moved dispatch logic into a truffle interop library. Started writing the overloads for UnresolvedConversions. We now statically generate UnresolvedConversions. These tweaks were mostly necessary to comply with out handling of extension methods and some previously-unknown-to-me dispatch rules. It should be finished by 2022-01-16.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). MethodDispatchLibrary needs implementations for each type. Starting that and then probably attending the (start of) our planning meeting.

**Additional Information:** Tuesday I'll pick up with Marcin implementing resolveFor for UnresolvedConversions and an 'invokeConversionNode' overload as it leans heavily on the caching bits. (Enso Bot - Jan 7, 2022)
<hr />
**Edward Kmett** reports a new **STANDUP** for today (2022-01-10):

**Today:** Long meeting is long. It should be finished by 2022-01-16.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Team meeting & candidate interview. MethodDispatchLibrary still needs implementations for each type.

**Additional Information:** My original plan was that I'll pick up with Marcin implementing resolveFor for UnresolvedConversions and an 'invokeConversionNode' overload as it leans heavily on the caching bits on Tuesday, but with the interview for Hubert and the team meeting it isn't clear how much of the day will be left after it is carved into pieces. Will balance concerns as I can. (Enso Bot - Jan 10, 2022)
<hr />
**Marcin Kostrzewa** reports a new **STANDUP** for today (2022-01-11):

**Today:** pair programming session with Edward, finishing the implementation of dispatch libraries for Enso builtin types It should be finished by 2022-01-16.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). a pair programming session with Edward, implementing conversion lookup logic and connecting it into the function call implementation (Enso Bot - Jan 11, 2022)
<hr />
**Edward Kmett** reports a new **STANDUP** for today (2022-01-11):

**Today:** Interviewed Hubert. Paired with Marcin. Implemented MethodDispatchLibrary implementations for all types. Epic organization meeting for "Most Important Libraries" It should be finished by 2022-01-16.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Meeting wdanilo. Implementing resolveFor for UnresolvedConversions. Inserting UnresolvedConversions into the syntax tree. Implementing invokeConversionNode. Throwing the switch on the mad science rig. Raising Frankenstein's monster into the sky. Waiting for lightning. Shouting "IT'S ALIVE!" (Enso Bot - Jan 12, 2022)
<hr />
**Marcin Kostrzewa** reports a new **STANDUP** for today (2022-01-12):

**Today:** pair programming session with Edward, getting the first version of the feature to work, still rough around the edges It should be finished by 2022-01-16.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). building a comprehensive test suite for the functionality and fixing all the missing bits (in error reporting and polyglot handling) (Enso Bot - Jan 12, 2022)
<hr />
**Edward Kmett** reports a new **STANDUP** for today (2022-01-12):

**Today:** IT'S ALIVE! In a branch anyways. A fair bit of cleanup is due but the main codepath works! It should be finished by 2022-01-16.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Test suite additions to ensure that all the expected parts work as intended. (There are maybe a dozen or so instances of the MethodDispatchLibrary that all had special logic added to handle them properly. We need to test these.)Then someone can QA to make sure any unexpected parts also work as you might expect. (Enso Bot - Jan 12, 2022)
<hr />
**Edward Kmett** reports a new **STANDUP** for today (2022-01-13):

**Today:** test suite for conversions added. we found a bug in one of the later compiler passes that swallows an error we place in the IR when the conversion is ill formed, but still have to resolve it, but well-formed conversions work now in the branch. It should be finished by 2022-01-16.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Running this last error swallowing collect {} to ground then rebasing and packaging up the merge request (Enso Bot - Jan 13, 2022)
<hr />
**Marcin Kostrzewa** reports a new **STANDUP** for today (2022-01-13):

**Today:** pair programming session with Edward. Implemented the full test suite and made all the errors user-friendly. In the process we've discovered that error reporting for top-level bindings was broken. Fixed the issue, but now it turns out that Enso libraries are full of top-level errors that were previously ignored. It should be finished by 2022-01-16.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Fixing the errors in standard library, then probably finishing the functionality and PR-ing it. (Enso Bot - Jan 13, 2022)
<hr />
**Marcin Kostrzewa** reports a new **STANDUP** for today (2022-01-14):

**Today:** pair programming session with Edward. Resolved all outstanding issues, got the functionality to a PR-able state and opened a PR. Then had a brainstorm session about sum types. It should be finished by 2022-01-16.

**Next Day:** Next day I will be working on the [##180901472 task](https://www.pivotaltracker.com/story/show/180901472). Doing some implementation work on the dataflow warnings, if Wojciech approves the design by my next work session. (Enso Bot - Jan 14, 2022)
<hr />
**Edward Kmett** reports a new **STANDUP** for today (2022-01-14):

**Today:** Pull request sent for https://github.com/enso-org/enso/pull/3227 Fixed the unrelated side issue with the failure of most type signatures to fail to parse. This feature now passes all of our tests. It should be finished by 2022-01-16.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180539669). Shepherding any last minute pull request details as I'm not terribly familiar with the arcane barriers the merge bot throws up here. Transitioning to studying how we could even try to implement sane sum types (issue #?)

**Additional Information:** Need to know the issue number for the sum type design task. (Enso Bot - Jan 14, 2022)
<hr />
**Edward Kmett** reports a new **STANDUP** for yesterday (2022-02-01):

**Today:** Split off a sub-issue to resolve the last tail-call issue with the PR for conversions, which is a minor perf improvement and shouldn't block. It should be finished by 2022-02-02.

**Next Day:** Next day I will be working on the [##180956811 task](https://www.pivotaltracker.com/story/show/180956811). Back to working on the type system with wdanilo and marcin. (Enso Bot - Feb 2, 2022)
<hr />
